gobblin-yarn/src/main/java/org/apache/gobblin/yarn/AbstractAppSecurityManager.java

```diff
@@ -0,0 +1,205 @@
+        String.format("Scheduling the login task with an interval of %d minute(s)", this.loginIntervalInMinutes));
+
+    // Schedule the Kerberos re-login task
+    this.loginExecutor.scheduleAtFixedRate(new Runnable() {
+      @Override
+      public void run() {
+        try {
+          loginAndScheduleTokenRenewal();
+        }catch(Exception e){
+          LOGGER.error("Error during login, will continue the thread and try next time.");
+        }
+      }
+    }, this.loginIntervalInMinutes, this.loginIntervalInMinutes, TimeUnit.MINUTES);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOGGER.info("Stopping the " + this.getClass().getSimpleName());
+
+    if (this.scheduledTokenRenewTask.isPresent()) {
+      this.scheduledTokenRenewTask.get().cancel(true);
+    }
+    ExecutorsUtils.shutdownExecutorService(this.loginExecutor, Optional.of(LOGGER));
+    ExecutorsUtils.shutdownExecutorService(this.tokenRenewExecutor, Optional.of(LOGGER));
+  }
+
+  protected void scheduleTokenRenewTask() {
+    LOGGER.info(String.format("Scheduling the token renew task with an interval of %d minute(s)",
+        this.tokenRenewIntervalInMinutes));
+
+    this.scheduledTokenRenewTask = Optional.<ScheduledFuture<?>>of(
+        this.tokenRenewExecutor.scheduleAtFixedRate(new Runnable() {
+          @Override
+          public void run() {
+            try {
+              renewDelegationToken();
+            } catch (IOException ioe) {
+              LOGGER.error("Failed to renew delegation token", ioe);
+              throw Throwables.propagate(ioe);
+            } catch (InterruptedException ie) {
+              LOGGER.error("Token renew task has been interrupted");
+              Thread.currentThread().interrupt();
+            }
+          }
+        }, this.tokenRenewIntervalInMinutes, this.tokenRenewIntervalInMinutes, TimeUnit.MINUTES));
+  }
+
+  //The whole logic for each re-login
+  public void loginAndScheduleTokenRenewal() {
+    try {
+      // Cancel the currently scheduled token renew task
+      if (scheduledTokenRenewTask.isPresent() && scheduledTokenRenewTask.get().cancel(true)) {
+        LOGGER.info("Cancelled the token renew task");
+      }
+
+      login();
+      if (firstLogin) {
+        firstLogin = false;
+      }
+
+      // Re-schedule the token renew task after re-login
+      scheduleTokenRenewTask();
+    } catch (IOException | InterruptedException ioe) {
+      LOGGER.error("Failed to login from keytab", ioe);
+      throw Throwables.propagate(ioe);
+    }
+  }
+
+  /**
+   * Write the current credentials to the token file.
+   */
+  protected synchronized void writeDelegationTokenToFile(Credentials cred) throws IOException {
+
+    if (this.fs.exists(this.tokenFilePath)) {
+      LOGGER.info("Deleting existing token file " + this.tokenFilePath);
+      this.fs.delete(this.tokenFilePath, false);
+    }
+
+    LOGGER.debug("creating new token file {} with 644 permission.", this.tokenFilePath);
```

Comment:
Good catch that this debug statement is definitely misleading. Internally, I see that there are references to 644, so perhaps that is where this is hard coded from. But in this class it's definitely changing the token file path to a 600 instead of a 644